### PR TITLE
[refactor] deviantart - improve results and clean up source code

### PR DESCRIPTION
**Pull request by return42**
## What does this PR do?

Devian's request and response forms has been changed.

- fixed title
- fixed time_range_dict to 'popular-*-***'
- use image from <noscript> if exists
- drop obsolete "http to https, remove domain sharding"
- use query URL https://www.deviantart.com/search/deviations?page=5&q=foo
- add searx/engines/deviantart.py to pylint check (test.pylint)

Error pattern::

    There DEBUG:searx:result: invalid title: {'url': 'https://www.deviantart.com/  ...

## Why is this change important?

Old one throws out no results but errors 

## How to test this PR locally?

Query:  `!da  millennium falcon`

### CODE REVIEW

File: searx/engines/deviantart.py [hunk 1]
```diff
@@ -7,76 +7,76 @@
...
-# get response from search-request
 def response(resp):
+    """get response from search-request"""
+
     results = []
 
-    # return empty array if a redirection code is returned
-    if resp.status_code == 302:
-        return []
+    with open("/tmp/dump.html", "w") as f:
```
Comment (kvch) [inline searx/engines/deviantart.py]: Leftover from debugging?

File: searx/engines/deviantart.py [hunk 2]
```diff
@@ -7,76 +7,76 @@
...
+            else:
+                img_tag = a_tag.xpath('.//img')
+            if not img_tag:
+                continue
+            img_tag = img_tag[0]
+
+            item = dict()
+
+            item['template'] = 'images.html'
+            item['url']      = a_tag.attrib.get('href')
```
Comment (kvch) [inline searx/engines/deviantart.py]: Why do these lines pass on pep8/pylint?

Why don't you follow the convention in python and searx where we set dict values like this:

```python
results.append({'url': url,
                'title': title,
                'img_src': img_src,
                'thumbnail_src': thumbnail_src,
                'template': 'images.html'})
```

File: searx/engines/deviantart.py [hunk 3]
```diff
@@ -7,76 +7,76 @@
...
 def request(query, params):
-    if params['time_range'] and params['time_range'] not in time_range_dict:
-        return params
+    """do search-request"""
 
-    params['url'] = search_url.format(page=params['pageno'],
-                                      query=urlencode({'q': query}))
+    # https://www.deviantart.com/search/deviations?page=5&q=foo
+
+    query =  dict(
```
Comment (kvch) [inline searx/engines/deviantart.py]: This indentation also seems weird. Why aren't you using the conventional indentation?

File: searx/engines/deviantart.py [hunk 4]
```diff
@@ -7,76 +7,76 @@
...
+)
 
+# search-url
+base_url = 'https://www.deviantart.com'
 
-# do search-request
 def request(query, params):
-    if params['time_range'] and params['time_range'] not in time_range_dict:
-        return params
+    """do search-request"""
```
Comment (kvch) [inline searx/engines/deviantart.py]: This comment seems unnecessary.

File: searx/engines/deviantart.py [hunk 5]
```diff
@@ -7,76 +7,76 @@
...
-base_url = 'https://www.deviantart.com/'
-search_url = base_url + 'search?page={page}&{query}'
-time_range_url = '&order={range}'
-
-time_range_dict = {'day': 11,
-                   'week': 14,
-                   'month': 15}
+time_range_dict = dict(
+    day     = 'popular-24-hours'
+    , week  = 'popular-1-week'
```
Comment (kvch) [inline searx/engines/deviantart.py]: Why are you putting the commas to the beginning of the line? We do not use this kind of format in searx.

**Review by kvch — CHANGES_REQUESTED** [review]
The fix look good, however, the indentation is comletely different from the convetions we use in searx. I am not sure why our style checker accepts this format. But we definitely have to enforce the style used in searx.

**Comment by return42** [comment]
@kvch thanks for having an eye on.

>  the indentation is comletely different from the convetions we use in searx.

I rebased and force pushed a patch.  

> Why aren't you using the conventional indentation?

[For new code pep8 suggests Donald E. Knuth's style](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator).

**Comment by kvch** [comment]
> > Why aren't you using the conventional indentation?
> 
> [For new code pep8 suggests Donald E. Knuth's style](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator).

I agree with those guidelines and use it where possible. However, it only applies to binary operators. Comma is not a binary operator. Ref: https://www.cs.auckland.ac.nz/references/unix/digital/AQTLTBTE/DOCU_062.HTM

File: searx/engines/deviantart.py [hunk 6]
```diff
@@ -7,76 +7,70 @@
...
+
+            if noscript_tag:
+                img_tag = noscript_tag[0].xpath('.//img')
+            else:
+                img_tag = a_tag.xpath('.//img')
+            if not img_tag:
+                continue
+            img_tag = img_tag[0]
+
+            results.append({
```
Comment (kvch) [inline searx/engines/deviantart.py:69]: Why have you changed the indentation here?

**Comment by return42** [comment]
>  I agree with those guidelines ... Comma is not a binary operator. 

.. this is not correct. The definition of a binary operation is very simple: A binary operation is a operation that combines two elements (called operands) to produce another element.

    a = b + c
    a = [b, c] 

In python and many other languages, a comma is the operator to concatenate two (or more) elements into a list (or tuple).

But this is only a academical view, the question is; what is common used (in searx) ..

> the indentation is comletely different from the convetions we use in searx

so I fixed to meet searx needs.

File: searx/engines/deviantart.py (hunk 6 above)
Comment (return42) [reply searx/engines/deviantart.py:69]: This is a `[refactor]` patch. So I decided to switch from the [_"Aligned with opening delimiter"_](https://www.python.org/dev/peps/pep-0008/#indentation) style to the [ _"Add 4 spaces (an extra level of indentation) to distinguish arguments from the rest"_](https://www.python.org/dev/peps/pep-0008/#indentation) style.

With an eye on the [Maximum Line Length](https://www.python.org/dev/peps/pep-0008/#maximum-line-length) I've got used to add an extra level of indentation (the latter style).

IMO: The growing diff is a common drawback of the patch, when switching from one code style to another. But I think this additional diff is acceptable in a *rewrite/refactoring* patch.

File: searx/engines/deviantart.py (hunk 4 above)
Comment (dalf) [reply searx/engines/deviantart.py]: pylint in the current configuration requires a docstring for each function, module and class.

`disable=bad-whitespace, duplicate-code, missing-function-docstring, missing-class-docstring, missing-module-docstring`
in `.pylintrc` remove this need.

Comment (return42) [reply searx/engines/deviantart.py]: > in .pylintrc remove this need.

I vote to disable only where it is needed to be disabled. E.g. I added

````python
# pylint: disable=missing-function-docstring
````

on top of the py-file.

**Comment by kvch** [comment]
> > I agree with those guidelines ... Comma is not a binary operator.
> 
> .. this is not correct. The definition of a binary operation is very simple: A binary operation is a operation that combines two elements (called operands) to produce another element.
> 
> ```
> a = b + c
> a = [b, c] 
> ```
> 
> In python and many other languages, a comma is the operator to concatenate two (or more) elements into a list (or tuple).
> 
> But this is only a academical view, the question is; what is common used (in searx) ..
> 

The academic view is not this. Binary means something has two parts. Binary operation is the operation that combines *exactly two* things:
__Let  S  be a non-empty set, and  ⋆  said to be a binary operation on  S  , if  a⋆b  is defined for all  a,b∈S . In other words,  ⋆  is a rule for any two elements in the set  S.__

Binary operators e.g. `+` can be rewritten like this `+(a, b)`. But in programming we use the format `a + b`. In your example `[a, b]` comma is not the operator. The operator is `list` and can be called like `list(a, b)`. But as Python comes with lots of syntactical sugar including `[` and `]` to represent lists, we do not have to write out the operator. But still it is just an "alias" for `list`.

File: searx/engines/deviantart.py (hunk 6 above)
Comment (kvch) [reply searx/engines/deviantart.py:69]: I am sorry if I was not clear. Personally I prefer this indentation, as if you add a new element to the end of the line, you do not have to add extra comma, thus decreasing the size of the diff. By changing the indentation I meant why did you put `:` at the same position in every line? 

If you add a new key which is longer than any of the existing keys, the whole dictionary has to be realigned, so the `:`s can be at the same vertical position again.

But I would be happy with this format:

```python
results.append({
    'template': 'images.html',
    'url': a_tag.attrib.get('href'),
    'img_src': img_tag.attrib.get('src'),
    'title': img_tag.attrib.get('alt'),
})
```

Comment (return42) [reply searx/engines/deviantart.py:69]: > I meant why did you put : at the same position in every line?

It is a (uneasy) compromise I often use for readability reasons.   Its a weak example, but in long lines and more key/value pairs, it has its value:

````python
{
    'img_src': img_tag.attrib.get('src'),
    'title': img_tag.attrib.get('alt'),
}
````

IMO: aligned its more _eye catching_ that merely different attributes from the same element are used as values

````python
{
    'img_src': img_tag.attrib.get('src'),
    'title':   img_tag.attrib.get('alt'),
}
````

The big drawback of is ..

> If you add a new key which is longer 

Yes, hoping there is enough space left does not scale very good in the diff.

I rebased my branch, removed the alignment (also in the `time_range_dict`)  and force-pushed.

**Comment by return42** [comment]
**update**: [commas are not operators](https://github.com/python/peps/issues/1706)

@kvch thanks for correcting me!

---

TL:DR;

> The academic view is not this. 
> But as Python comes with lots of syntactical sugar including 

In C exists [the comma operator (binary operator)](https://en.wikipedia.org/wiki/Comma_operator) but I think this one is off topic here. I refer this one here, just to distinguish "binary operator" in between a mathematical context and a programming context.

> Binary means something has two parts.  ....

I have to admit that I've never thought about it much.  I see the point in your explanation (thanks) and now I have a doubt how to read _"binary operators"_ in this sentence:

[For decades the recommended style was to break after binary operators. But this can hurt readability in two ways: the operators tend to get scattered across different columns on the screen, and each operator is moved away from its operand and onto the previous line. ](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator)

Since this section is about _readability_  I didn't distinguished between a binary operator in mathematics and a comma in a programming context (e.g. to separating the items of an argument list).  May be my fail .. 

With your argumentation in mind, I don't have any longer a strong opinion.

**update**: I added an issue to the python/peps repo ["PEP-8: Should a line break before or after a comma?"](https://github.com/python/peps/issues/1706)

**Comment by kvch** [comment]
> **update**: [commas are not operators](https://github.com/python/peps/issues/1706)
> 
> @kvch thanks for correcting me!
> 
> TL:DR;
> 
> > The academic view is not this.
> > But as Python comes with lots of syntactical sugar including
> 
> In C exists [the comma operator (binary operator)](https://en.wikipedia.org/wiki/Comma_operator) but I think this one is off topic here. I refer this one here, just to distinguish "binary operator" in between a mathematical context and a programming context.

I was not aware of this operator in C. I had coded in C for a few years, but for some reason never encountered it. I learnt something new today. Thanks for that. :)

**Review by kvch — DISMISSED** [review]

**Comment by return42** [comment]
> I learnt something new today. Thanks for that. :)

Same on my side. Thanks for the discussion that helps me to correct a long standing mistake!

I rebased and force pushed .. if you have time to merge .. afterwards I will fix [2285](https://github.com/searx/searx/pull/2285#issuecomment-720967342) according to your remarks from this PR.

**Comment by return42** [comment]
Sorry for being impatient / this engine is already approved by @kvch, can we please merge so I can continue with the next engine (#2285) / thanks!

**Comment by dalf** [comment]
Thank you !
